python/cuml/dask/datasets/make_regression.py

```diff
@@ -0,0 +1,216 @@
+#
+# Copyright (c) 2019-2020, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import dask.array as da
+import cupy as cp
+
+
+def make_low_rank_matrix(n_samples=100, n_features=100, effective_rank=10,
+                         tail_strength=0.5, random_state=None, n_parts=1):
+    """ Generate a mostly low rank matrix with bell-shaped singular values
+
+    Parameters
+    ----------
+    n_samples : int, optional (default=100)
+        The number of samples.
+    n_features : int, optional (default=100)
+        The number of features.
+    effective_rank : int, optional (default=10)
+        The approximate number of singular vectors required to explain most of
+        the data by linear combinations.
+    tail_strength : float between 0.0 and 1.0, optional (default=0.5)
+        The relative importance of the fat noisy tail of the singular values
+        profile.
+    random_state : int
+        Determines random number generation for dataset creation. Pass an int
+        for reproducible output across multiple function calls.
+    n_parts : int, optional (default=1)
+        The number of parts of work.
+
+    Returns
+    -------
+    X : Dask-cuPy array of shape [n_samples, n_features]
+        The matrix.
+    """
+    rs = da.random.RandomState(seed=random_state,
+                               RandomState=cp.random.RandomState)
+    n = min(n_samples, n_features)
+
+    def generate_chunks_for_qr(total_size, min_size, n_parts):
+
+        n_total_per_part = max(1, int(total_size / n_parts))
+        if n_total_per_part > min_size:
+            min_size = n_total_per_part
+
+        n_partitions = int(max(1, total_size / min_size))
+        rest = total_size % (n_partitions * min_size)
+        chunks_list = [min_size for i in range(n_partitions-1)]
+        chunks_list.append(min_size + rest)
+        return tuple(chunks_list)
+
+    # Random (ortho normal) vectors
+    m1 = rs.standard_normal((n_samples, n),
+                            chunks=(generate_chunks_for_qr(n_samples,
+                                                           n, n_parts), -1))
+    u, _ = da.linalg.qr(m1)
+
+    m2 = rs.standard_normal((n, n_features),
+                            chunks=(-1, generate_chunks_for_qr(n_features,
+                                                               n, n_parts)))
+    v, _ = da.linalg.qr(m2)
```

Comment:
Have you verified this is using cupy end-to-end (eg. not creating any arrays on host internally?) In general the Dask.Array does a good job at trying to rely on the underlying ndarray implementation to perform computations, but I've encountered some places where they are using numpy explicitly, which has a massive impact on performance.